pkg/action/executor/move_util.go

```diff
@@ -0,0 +1,551 @@
+	"time"
+
+	"github.com/golang/glog"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kclient "k8s.io/client-go/kubernetes"
+	api "k8s.io/client-go/pkg/api/v1"
+)
+
+//compare two version strings
+func compareVersion(version1, version2 string) int {
+	a1 := strings.Split(version1, ".")
+	a2 := strings.Split(version2, ".")
+
+	l1 := len(a1)
+	l2 := len(a2)
+	mlen := l1
+	if mlen < l2 {
+		mlen = l2
+	}
+
+	for i := 0; i < mlen; i++ {
+		b1 := 0
+		if i < l1 {
+			if tmp, err := strconv.Atoi(a1[i]); err == nil {
+				b1 = tmp
+			}
+		}
+
+		b2 := 0
+		if i < l2 {
+			if tmp, err := strconv.Atoi(a2[i]); err == nil {
+				b2 = tmp
+			}
+		}
+
+		if b1 != b2 {
+			return b1 - b2
+		}
+	}
+
+	return 0
+}
+
+//retry to execute a function with a timeout
+func retryDuring(attempts int, timeout time.Duration, sleep time.Duration, myfunc func() error) error {
+	t0 := time.Now()
+
+	var err error
+	for i := 0; ; i++ {
+		if err = myfunc(); err == nil {
+			glog.V(3).Infof("[retry-%d/%d] success", i+1, attempts)
+			return nil
+		}
+
+		glog.V(2).Infof("[retry-%d/%d] Warning %s", i+1, attempts, err.Error())
+		if i >= (attempts - 1) {
+			break
+		}
+
+		if timeout > 0 {
+			if delta := time.Now().Sub(t0); delta > timeout {
+				err = fmt.Errorf("after %d attepmts (during %d) last error: %s", i, delta, err.Error())
+				glog.Error(err.Error())
+				return err
+			}
+		}
+
+		if sleep > 0 {
+			time.Sleep(sleep)
+		}
+	}
+
+	err = fmt.Errorf("after %d attepmts last error: %s", attempts, err.Error())
+	glog.Error(err.Error())
+	return err
+}
+
+// concate errors
+func addErrors(prefix string, err1, err2 error) error {
+	if err1 == nil && err2 == nil {
+		return nil
+	}
+
+	rerr := fmt.Errorf("%v ", prefix)
+	if err1 != nil {
+		rerr = fmt.Errorf("%v %v", rerr.Error(), err1.Error())
+	}
+
+	if err2 != nil {
+		rerr = fmt.Errorf("%v %v", rerr.Error(), err2.Error())
+	}
+
+	return rerr
+}
+
+//get kind and name of pod's parent Controller
+func getParentInfo(pod *api.Pod) (string, string, error) {
+	//1. check ownerReferences:
+	if pod.OwnerReferences != nil && len(pod.OwnerReferences) > 0 {
+		for _, owner := range pod.OwnerReferences {
+			if *owner.Controller {
+				return owner.Kind, owner.Name, nil
+			}
+		}
+	}
+
+	glog.V(4).Infof("cannot find pod-%v/%v parent by OwnerReferences.", pod.Namespace, pod.Name)
+
+	//2. check annotations:
+	if pod.Annotations != nil && len(pod.Annotations) > 0 {
+		key := "kubernetes.io/created-by"
+		if value, ok := pod.Annotations[key]; ok {
+
+			var ref api.SerializedReference
+
+			if err := json.Unmarshal([]byte(value), &ref); err != nil {
+				err = fmt.Errorf("failed to decode parent annoation:%v", err.Error())
+				glog.Errorf("%v\n%v", err.Error(), value)
+				return "", "", err
+			}
+
+			return ref.Reference.Kind, ref.Reference.Name, nil
+		}
+	}
+
+	glog.V(4).Infof("cannot find pod-%v/%v parent by Annotations.", pod.Namespace, pod.Name)
+
+	return "", "", nil
+}
+
+func copyPodInfo(oldPod, newPod *api.Pod) {
```

Comment:
I cannot explain why some fields are copied, some are not.
I tested them. It works now. I will keep a mind about this if there is some misbehavior of copied pod.